From e39f893b6d5d619c4c1d5d635794bbdcb322c5dc Mon Sep 17 00:00:00 2001
From: liusheng <liusheng@eswincomputing.com>
Date: Thu, 26 Sep 2024 17:20:56 +0800
Subject: [PATCH 273/652] WIN2030-16139:fix:optimize hae blit with stall on CPU
 issue

Changelogs:
1. optimize hae blit with stall on CPU issue
2. optimize user memory import for pfn type
Change-Id: I1bc49cabda603832afde4c611724149c815cbcd9
Signed-off-by: liusheng <liusheng@eswincomputing.com>
---
 .../hal/kernel/arch/gc_hal_kernel_hardware.c  |  4 ++-
 .../arch/gc_hal_kernel_hardware_waitlink_fe.c |  3 +-
 .../gc_hal_kernel_allocator_user_memory.c     | 34 +++++++++++++------
 3 files changed, 27 insertions(+), 14 deletions(-)

diff --git a/drivers/staging/media/eswin/hae/hal/kernel/arch/gc_hal_kernel_hardware.c b/drivers/staging/media/eswin/hae/hal/kernel/arch/gc_hal_kernel_hardware.c
index bebbbd4a7edf..0cd8c0ab4dae 100644
--- a/drivers/staging/media/eswin/hae/hal/kernel/arch/gc_hal_kernel_hardware.c
+++ b/drivers/staging/media/eswin/hae/hal/kernel/arch/gc_hal_kernel_hardware.c
@@ -2033,7 +2033,8 @@ gckHARDWARE_Construct(gckOS Os, gckKERNEL Kernel, gckHARDWARE *Hardware)
     /* Check if big endian */
     hardware->bigEndian = (*(gctUINT8 *)&data == 0xff);
 
-    gcmkONERROR(gckOS_CreateSignal(Os, gcvTRUE, &hardware->feIdleSignal));
+    gcmkONERROR(gckOS_CreateSignal(Os, gcvFALSE, &hardware->feIdleSignal));
+    gcmkONERROR(gckOS_Signal(Os, hardware->feIdleSignal, gcvTRUE));
 
     /* Initialize the fast clear. */
     gcmkONERROR(gckHARDWARE_SetFastClear(hardware, -1, -1));
@@ -9664,6 +9665,7 @@ gckHARDWARE_ExecuteFunctions(gcsFUNCTION_EXECUTION_PTR Execution)
             gcmkONERROR(gckWLFE_Execute(hardware, address,
                                         Execution->funcCmd[i].bytes));
             gcmkONERROR(gckOS_WaitSignal(hardware->os, hardware->feIdleSignal, gcvFALSE, gcdGPU_2D_TIMEOUT));
+            gcmkONERROR(gckOS_Signal(hardware->os, hardware->feIdleSignal, gcvTRUE));
         }
 
 #if gcdLINK_QUEUE_SIZE
diff --git a/drivers/staging/media/eswin/hae/hal/kernel/arch/gc_hal_kernel_hardware_waitlink_fe.c b/drivers/staging/media/eswin/hae/hal/kernel/arch/gc_hal_kernel_hardware_waitlink_fe.c
index bb92f5f58e0a..52907911bf20 100644
--- a/drivers/staging/media/eswin/hae/hal/kernel/arch/gc_hal_kernel_hardware_waitlink_fe.c
+++ b/drivers/staging/media/eswin/hae/hal/kernel/arch/gc_hal_kernel_hardware_waitlink_fe.c
@@ -1130,7 +1130,7 @@ gckWLFE_Execute(gckHARDWARE Hardware, gctADDRESS Address, gctUINT32 Bytes)
             gcmkVERIFY_OK(gckOS_ReadRegisterEx(Hardware->os, Hardware->kernel,
                                                0x00004, &idle));
             if (idle != 0x7FFFFFFF) {
-                gcmkONERROR(gckOS_WaitSignal(Hardware->os, Hardware->feIdleSignal, gcvFALSE, gcdGPU_2D_TIMEOUT));
+                gcmkONERROR(gckOS_WaitSignal(Hardware->os, Hardware->feIdleSignal, gcvTRUE, gcdGPU_2D_TIMEOUT));
             }
         } while (idle != 0x7FFFFFFF);
     }
@@ -1168,7 +1168,6 @@ gckWLFE_Execute(gckHARDWARE Hardware, gctADDRESS Address, gctUINT32 Bytes)
     gcmkONERROR(gckOS_MemoryBarrier(Hardware->os, gcvNULL));
 
     if (Hardware->type == gcvHARDWARE_2D) {
-        gcmkONERROR(gckOS_Signal(Hardware->os, Hardware->feIdleSignal, gcvFALSE));
         gcmkONERROR(gckOS_WriteRegisterEx(Hardware->os, Hardware->kernel,
                                           0x00658, control));
     } else {
diff --git a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
index 151e006da98f..84734b7cbe81 100644
--- a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
+++ b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/allocator/default/gc_hal_kernel_allocator_user_memory.c
@@ -334,6 +334,7 @@ import_pfn_map(gckOS Os, struct device *dev, struct um_desc *um,
     int result = 0;
     size_t pageCount = 0;
 	unsigned int           data      = 0;
+
     if (!current->mm)
         return -ENOTTY;
 
@@ -370,24 +371,27 @@ import_pfn_map(gckOS Os, struct device *dev, struct um_desc *um,
         int ret = 0;
         ret = follow_pfn(vma, addr, &pfns[i]);
         if (ret < 0) {
+            /* Case maybe provides unmapped addr. */
             ret = gckOS_ReadMappedPointer(Os, (gctPOINTER)addr, &data);
             if (!ret)
                 ret = follow_pfn(vma, addr, &pfns[i]);
+
             if (ret < 0) {
-            up_read(&current_mm_mmap_sem);
-            goto err;
+                up_read(&current_mm_mmap_sem);
+                goto err;
             }
         }
 #else
         /* protect pfns[i] */
         spinlock_t  *ptl;
-        pgd_t *pgd;
+        pgd_t       *pgd;
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 9, 0)
-        p4d_t *p4d;
-# endif
-        pud_t *pud;
-        pmd_t *pmd;
-        pte_t *pte;
+        p4d_t       *p4d;
+#    endif
+        pud_t       *pud;
+        pmd_t       *pmd;
+        pte_t       *pte;
+
         pgd = pgd_offset(current->mm, addr);
         if (pgd_none(*pgd) || pgd_bad(*pgd))
             goto err;
@@ -419,14 +423,22 @@ import_pfn_map(gckOS Os, struct device *dev, struct um_desc *um,
             if (gcmIS_SUCCESS(gckOS_ReadMappedPointer(Os, (gctPOINTER)addr, &data)))
                 pte = pte_offset_map_lock(current->mm, pmd, addr, &ptl);
         if (!pte_present(*pte)) {
-            pte_unmap_unlock(pte, ptl);
-            goto err;
+            if (pte)
+                pte_unmap_unlock(pte, ptl);
+
+            /* Case maybe provides unmapped addr. */
+            if (gcmIS_SUCCESS(gckOS_ReadMappedPointer(Os, (gctPOINTER)addr, &data)))
+                pte = pte_offset_map_lock(current->mm, pmd, addr, &ptl);
+
+            if (!pte_present(*pte)) {
+                pte_unmap_unlock(pte, ptl);
+                goto err;
+            }
         }
         }
         pfns[i] = pte_pfn(*pte);
         pte_unmap_unlock(pte, ptl);
 #endif
-
         /* Advance to next. */
         addr += PAGE_SIZE;
     }
-- 
2.54.0

